Adding third party TRITON_PLUGIN_DIRS is broken

### Describe the bug

When using new version of triton, and trying to build triton_shared with TRITON_PLUGIN_DIRS=$pwd/triton_shared, I get the error:
```
error: Error: setup script specifies an absolute path:
/home/runner/work/triton-shared/triton-shared/triton_shared/backend
setup() arguments must *always* be /-separated paths relative to the
setup.py directory, *never* absolute paths.
```
This happens because in https://github.com/triton-lang/triton/pull/6381 there was major change to setup.py, where previously absolute paths were converted to relative paths (which is what the setup command requires), but that is no longer happening.

Forcing TRITON_PLUGIN_DIRS to all be relative paths does sometimes resolve this, but it doesn't work very well, because the paths will first be interpreted as the relative to the working directory of the python process running setup.py and then the relative path to setup.py itself, and this is not always going to be the same (e.g. pip install sets cwd to be a temp directory).

### Environment details

Triton commit: fbfa627c3adadc8e83f56551150e0f138aaddd5a

`triton-shared` unfortunately fails to build for me:

```
/tmp/triton_shared/lib/Conversion/TritonToLinalgExperimental/ReconcilePtrCastsPass.cpp: In member function ‘virtual llvm::LogicalResult {anonymous}::FromMemrefConverter::matchAndRewrite(mlir::UnrealizedConversionCastOp, mlir::PatternRewriter&) const’:
/tmp/triton_shared/lib/Conversion/TritonToLinalgExperimental/ReconcilePtrCastsPass.cpp:97:42: error: no matching function for call to ‘mlir::ptr::PtrType::get(mlir::MLIRContext*) const’
   97 |           op->getLoc(), ptr::PtrType::get(rewriter.getContext()), rankedMemref);
      |                         ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
```

and more. But I think it ran enough logic to confirm the symlinking works.

Could you test #6627 to confirm that it fixes the issue for you?

--- a/MANIFEST.in
+++ b/MANIFEST.in
@@ -5,9 +5,7 @@ graft include
 graft lib
 graft python/src
 graft python/test
-graft python/triton/backends/amd
-graft python/triton/backends/nvidia
-graft python/triton/tools/extra/cuda
+graft python/triton
 graft test
 graft third_party
 graft unittest

--- a/setup.py
+++ b/setup.py
@@ -20,9 +20,18 @@ from setuptools import Extension, find_packages, setup
 from setuptools.command.build_ext import build_ext
 from setuptools.command.build_py import build_py
 from setuptools.command.develop import develop
+from setuptools.command.egg_info import egg_info
+from setuptools.command.install import install
+from setuptools.command.sdist import sdist
+
 from dataclasses import dataclass
 
 import pybind11
+
+try:
+    from setuptools.command.bdist_wheel import bdist_wheel
+except ImportError:
+    from wheel.bdist_wheel import bdist_wheel
 
 try:
     from setuptools.command.editable_wheel import editable_wheel
@@ -587,6 +596,10 @@ def get_package_dirs():
     yield ("", "python")
 
     for backend in backends:
+        # we use symlinks for external plugins
+        if backend.is_external:
+            continue
+
         yield (f"triton.backends.{backend.name}", backend.backend_dir)
 
         if backend.language_dir:
@@ -605,8 +618,11 @@ def get_package_dirs():
         yield ("triton.profiler", "third_party/proton/proton")
 
 
-def add_link_to_backends():
+def add_link_to_backends(external_only):
     for backend in backends:
+        if external_only and not backend.is_external:
+            continue
+
         update_symlink(backend.install_dir, backend.backend_dir)
 
         if backend.language_dir:
@@ -635,23 +651,53 @@ def add_link_to_proton():
     update_symlink(proton_install_dir, proton_dir)
 
 
-def add_links():
-    add_link_to_backends()
-    if check_env_flag("TRITON_BUILD_PROTON", "ON"):  # Default ON
+def add_links(external_only):
+    add_link_to_backends(external_only=external_only)
+    if not external_only and check_env_flag("TRITON_BUILD_PROTON", "ON"):  # Default ON
         add_link_to_proton()
+
+
+class plugin_bdist_wheel(bdist_wheel):
+
+    def run(self):
+        add_links(external_only=True)
+        super().run()
 
 
 class plugin_develop(develop):
 
     def run(self):
-        add_links()
+        add_links(external_only=False)
         super().run()
 
 
 class plugin_editable_wheel(editable_wheel):
 
     def run(self):
-        add_links()
+        add_links(external_only=False)
+        super().run()
+
+
+class plugin_egg_info(egg_info):
+
+    def run(self):
+        add_links(external_only=True)
+        super().run()
+
+
+class plugin_install(install):
+
+    def run(self):
+        add_links(external_only=True)
+        super().run()
+
+
+class plugin_sdist(sdist):
+
+    def run(self):
+        for backend in backends:
+            if backend.is_external:
+                raise RuntimeError("sdist cannot be used with TRITON_PLUGIN_DIRS")
         super().run()
 
 
@@ -713,11 +759,15 @@ setup(
     include_package_data=True,
     ext_modules=[CMakeExtension("triton", "triton/_C/")],
     cmdclass={
+        "bdist_wheel": plugin_bdist_wheel,
         "build_ext": CMakeBuild,
         "build_py": CMakeBuildPy,
         "clean": CMakeClean,
         "develop": plugin_develop,
         "editable_wheel": plugin_editable_wheel,
+        "egg_info": plugin_egg_info,
+        "install": plugin_install,
+        "sdist": plugin_sdist,
     },
     zip_safe=False,
     # for PyPI

--- a/setup.py
+++ b/setup.py
@@ -618,6 +618,28 @@ def get_package_dirs():
         yield ("triton.profiler", "third_party/proton/proton")
 
 
+def get_packages():
+    yield from find_packages(where="python")
+
+    for backend in backends:
+        yield f"triton.backends.{backend.name}"
+
+        if backend.language_dir:
+            # Install the contents of each backend's `language` directory into
+            # `triton.language.extra`.
+            for x in os.listdir(backend.language_dir):
+                yield f"triton.language.extra.{x}"
+
+        if backend.tools_dir:
+            # Install the contents of each backend's `tools` directory into
+            # `triton.tools.extra`.
+            for x in os.listdir(backend.tools_dir):
+                yield f"triton.tools.extra.{x}"
+
+    if check_env_flag("TRITON_BUILD_PROTON", "ON"):  # Default ON
+        yield "triton.profiler"
+
+
 def add_link_to_backends(external_only):
     for backend in backends:
         if external_only and not backend.is_external:
@@ -739,9 +761,6 @@ def get_git_version_suffix():
 # keep it separate for easy substitution
 TRITON_VERSION = "3.3.0" + get_git_version_suffix() + os.environ.get("TRITON_WHEEL_VERSION_SUFFIX", "")
 
-package_dirs = dict(get_package_dirs())
-extra_packages = [x for x in package_dirs if x != ""]
-
 setup(
     name=os.environ.get("TRITON_WHEEL_NAME", "triton"),
     version=TRITON_VERSION,
@@ -753,8 +772,8 @@ setup(
         "setuptools>=40.8.0",
         "importlib-metadata; python_version < '3.10'",
     ],
-    packages=find_packages(where="python") + extra_packages,
-    package_dir=package_dirs,
+    packages=list(get_packages()),
+    package_dir=dict(get_package_dirs()),
     entry_points=get_entry_points(),
     include_package_data=True,
     ext_modules=[CMakeExtension("triton", "triton/_C/")],

